doc/designs/hsm.md

```diff
@@ -0,0 +1,277 @@
+
+This would be an alternative, but unsupported, way to install IPA with a HSM. We could allow this but it would require parsing the config file to pull out the relevant information and it seems like a lot more effort, and error-prone, than adding the required options to the cli.
+
+This can be confusing because this same override needs to be passed to ipa-kra-install, ipa-replica-install and ipa-ca-install for installation to succeed. The chance of a user forgetting to do this are high.
+
+## Implementation
+
+### Known changes required
+#### IPA
+- Add new installation options for token, HSM library and password. Retrieve the options on replica installation and when adding a CA or KRA to an existing server (ipa-ca-install/ipa-kra-install).
+- Extend the expected tracking to include the token name
+- Depending on how certmonger addresses executing helpers, options will need to be provided to utilize it.
+
+#### certmonger
+- certmonger needs to setuid/setgid to pkiuser when performing operations to ensure filesystem permissions are retained for softhsm. https://pagure.io/certmonger/issue/243
+
+
+#### NSS
+- NSS requires a change to avoid prompting the user for the token PIN multiple times during installation, https://bugzilla.mozilla.org/show_bug.cgi?id=1782980
+
+#### ipa-healthcheck
+- ipa-healthcheck has no understanding at all of tokens https://github.com/freeipa/freeipa-healthcheck/issues/276
+
+#### pki
+- sub CA's are generated in the softokn with a strange naming convention: ```NSS Certificate DB:softhsm_token::caSigningCert cert-pki-ca UUID```. Needs investigation. Not deemed a show-stopper for MVP.
+
+
+### Dependencies
+
+HSM support for IPA relies on changes in multiple external projects:
+
+- pki 11.3.0 contains the minimum set of changes required
+- updated certmonger
+- updated ipa-healthcheck
+- updated NSS
+
+### SoftHSM2 token management
+
+To create a softhsm token:
+
+runuser -u pkiuser -- /usr/bin/softhsm2-util --init-token --free --pin password --so-pin TokenSOpassword --label softhsm_token
+
+The runuser is required so the resulting files are available to pki.
+
+To remove the token:
+
+softhsm2-util --delete-token --token softhsm_token
+
+Initial server install w/DNS
+
+ipa-server-install -a password -p dmpassword -r EXAMPLE.TEST -U --setup-dns --allow-zone-overlap --no-forwarders -N --auto-reverse --pki-config-override=/root/pki.ini
+
+## Feature Management
+
+### UI
+
+There should be no noticeable change.
+
+### CLI
+
+There should be no noticeable change.
+
+The HSM attributes will not be visible via the CLI except through ```ipa ca-show ipa --all```
+
+This information will be retrieved over LDAP rather than the API due to bootstrapping chicken-and-egg.
+
+
+## Test plan
+
+A subset of existing tests will be subclassed and executed with HSM enabled (softhsm) in a similar way that random serial number testing was done.
+
+## Troubleshooting and debugging
+
```

Comment:
Mention the ipaca object stored in LDAP, which contains the ipaCaHSMConfiguration attribute.

Are there specific commands to check the content of the HSM? Is certutil enough?

What is the behavior when the HSM is not reachable any more? Does PKI have specific error messages?

Reply:
1 and 2 done. For HSM not available, I don't know. Probably certificate not found. I haven't yet gotten my hands on a real HSM. I'll give that a go when I get one.